### PR TITLE
Lux AI 2022 v1.1.0

### PR DESCRIPTION
**Pull request by StoneT2000**

### CODE REVIEW

**Comment by StoneT2000** [comment]
@bovard

File: kaggle_environments/envs/lux_ai_2022/luxai2022/env.py [hunk 1]
```diff
@@ -211,23 +211,35 @@ def _handle_bid(self, actions):
...
+        if highest_bid_agent is None:
             # no valid bids made, all agents failed.
             pass
         else:
             self.state.teams[highest_bid_agent].init_water -= highest_bid
             self.state.teams[highest_bid_agent].init_metal -= highest_bid
-            self.state.teams[highest_bid_agent].factories_to_place += 1
+            self.state.teams[highest_bid_agent].place_first = True
         return failed_agents
     def _handle_factory_placement_step(self, actions):
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:227]: Players might want to bid to go 2nd. Either need to add a second val to the bid to show go first or second, or allow "negative" bid values which means "I want to go second" and validation that.

File: kaggle_environments/envs/lux_ai_2022/luxai2022/env.py [hunk 2]
```diff
@@ -211,23 +211,35 @@ def _handle_bid(self, actions):
...
-        elif highest_bid_agent is None:
+        if highest_bid_agent is None:
             # no valid bids made, all agents failed.
             pass
         else:
             self.state.teams[highest_bid_agent].init_water -= highest_bid
             self.state.teams[highest_bid_agent].init_metal -= highest_bid
-            self.state.teams[highest_bid_agent].factories_to_place += 1
+            self.state.teams[highest_bid_agent].place_first = True
         return failed_agents
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:226]: Players might want to bid to go 2nd. Either need to add a second val to the bid to show go first or second, or allow "negative" bid values which means "I want to go second" the player with highest absolute value wins and gets their desired placement position.

File: kaggle_environments/envs/lux_ai_2022/luxai2022/config.py [hunk 3]
```diff
@@ -121,8 +120,6 @@ class EnvConfig:
     NUM_WEATHER_EVENTS_RANGE: List[int] = dataclasses.field(default_factory=lambda:[3,5])
     WEATHER: dict = dataclasses.field(default_factory=lambda:dict(
         MARS_QUAKE=dict(
-            # amount of rubble generated under each robot per turn
-            RUBBLE=dict(LIGHT=1, HEAVY=10),
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/config.py:125]: Where did these go?

File: kaggle_environments/envs/lux_ai_2022/luxai2022/env.py [hunk 4]
```diff
@@ -341,17 +359,34 @@ def _handle_factory_build_actions(self, actions_by_type: ActionsByType, weather_
...
-                unit_type=factory_build_action.unit_type,
-                pos=factory.pos.pos,
-            )
+
+            # note that this happens after unit resource pickup. Thus we have to reverify these actions
+            # if self.state.env_cfg.validate_action_space:
+
             if factory_build_action.unit_type == UnitType.HEAVY:
+                if factory.cargo.metal < self.env_cfg.ROBOTS["HEAVY"].METAL_COST:
+                    self._log(f"{factory} doesn't have enough metal to build a heavy despite having enough metal at the start of the turn. This is likely because a unit picked up some of the metal.")
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:368]: unfortunate!

File: kaggle_environments/envs/lux_ai_2022/luxai2022/env.py [hunk 5]
```diff
@@ -644,8 +681,11 @@ def add_factory(self, team: Team, pos: np.ndarray):
         factory.cargo.water = self.env_cfg.INIT_WATER_METAL_PER_FACTORY
         factory.cargo.metal = self.env_cfg.INIT_WATER_METAL_PER_FACTORY
         factory.power = self.env_cfg.INIT_POWER_PER_FACTORY
-        if not self.state.board.spawn_masks[team.agent][pos[0], pos[1]]:
-            self._log(f"{team.agent} cannot place factory at {pos[0]}, {pos[1]} as it is on the other half of map.")
+        if self.state.board.valid_spawns_mask[pos[0], pos[1]] == 0:
+            # Check if any tiles under the factory are invalid spawn tile.
+            # TODO - min distance between Factories? stone: I think it's a bad strategy to try and enclose a opponent factory anyway,
```
Comment (bovard) [inline kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:686]: is there a min distance?

Comment (bovard) [reply kaggle_environments/envs/lux_ai_2022/luxai2022/env.py:686]: looks like we are doing below

**Review by bovard — CHANGES_REQUESTED** [review]

File: kaggle_environments/envs/lux_ai_2022/luxai2022/config.py (hunk 3 above)
Comment (StoneT2000) [reply kaggle_environments/envs/lux_ai_2022/luxai2022/config.py:125]: by default we use the values for rubble after destruction under the units tab (same number)

**Comment by StoneT2000** [comment]
added negative bids and repeat action fix

**Pull request merged**